### ISSUE TITLE
sort-comp rules does not work

### ISSUE THREAD

**Issue by TSMMark** [yannickcr/eslint-plugin-react#1490]
Hello, I am not getting warnings or errors regardless of the order of my React class method definitions.

I've tried with default config, and various other configs, but nothing will cause any kind of eslint failure. Every other of the many rules in my eslintrc works fine except this one: `react/sort-comp`. Here's the current config:

```json
"react/sort-comp": [2, {
  "order": [
    "type-annotations",
    "static-methods",
    "lifecycle",
    "/^on.+$/",
    "/^(get|set)(?!(InitialState$|DefaultProps$|ChildContext$)).+$/",
    "everything-else",
    "/^render.+$/",
    "render"
  ]
}],
```

Versions:
eslint@4.9.0
eslint-plugin-react@7.4.0

**Comment by OZZlE** [comment]
We're using the default configuration but it seems veeeeery buggy..

I added one non native function to a file, it asked me to move it to another place, so I moved it there, then it asks me to move another existing method as well.. and then you can continue like that forever, making unnecessary changes.

It seems to want to sort non static/native react functions by alphabet but fails spectacularly. If you work in a team where anyone can add a function in a branch and you merge two different branches you might get a breaking eslint error.. doesn't make any sense to me to sort by alphabet!

**Comment by ljharb** [comment]
@OZZlE it’s not that it goes forever, it’s that it can’t compute the optimal sorting and provide those as eslint warnings - all it can do is tell you it’s in the wrong place.

**Comment by ljharb** [comment]
@TSMMark can you provide code that can reproduce the problem (with this config)?

**Comment by TSMMark** [comment]
Oh geez this was so long ago I don't think we have any class components left — Wait I think I remember we used to use fat arrow/auto-binding instance method syntax sugar, like:

```js
class MyComponent {
  methodZ = () => {}
  methodA = () => {}
}
```

Not sure what the name for that syntax sugar is but AFAIK it's not accepted in ES yet so maybe that's why it didn't work with the lint rule? Not sure but feel free to close this issue if you want. Thanks for following up

**Comment by ljharb** [comment]
In fact class fields are stage 4, in ES2022.

It's likely that back in 2017, that the rule didn't support those; it does now.


<sub>altho, the far better pattern is this:</sub>
```js
methodZ = this.methodZ.bind(this);
methodZ() {}
```
<sub>… arrow functions should never be used with class fields.</sub>

I'll close it, thanks!

**Comment by TSMMark** [comment]
Good to know the terminology of "class fields", thanks!

> arrow functions should never be used with class fields.

Why is that? The syntax is so much more terse with arrows I gotta say

**Comment by ljharb** [comment]
Indeed it is - but the downside is that you create an entirely new unoptimized function for each instance - with the method i showed, you create one method on the prototype, shared by every instance and easily mocked/stubbed/spied for tests, and the only per-instance part is the bound method. The shared method still gets optimized.
